File: src/ctx/graph.py
from __future__ import annotations

import time
from pathlib import Path
from typing import Any

from .config import find_ctx
from .db import connect
from .search import apply_budget


def _row_hit(row: Any, kind: str, why: str, score: float = 1.0) -> dict[str, Any]:
    return {
        "path": row["path"], "start": row["start"], "end": row["end"],
        "symbol": row["name"], "kind": kind, "sig": row["sig"] or "",
        "snippet": row["snippet"] or "", "score": score, "why": why,
    }


def graph_query(op: str, symbol: str, *, depth: int = 2, budget_tokens: int = 1500,
                start: Path | str = ".") -> dict[str, Any]:
    started = time.perf_counter()
    conn = connect(find_ctx(start) / "index.sqlite")
    coverage = "complete"
    try:
        definitions = conn.execute(
            "SELECT s.*,f.path,f.is_test,f.is_vendor FROM symbols s JOIN files f ON f.id=s.file_id "
            "WHERE s.name=? OR s.qualname=? ORDER BY f.is_test,f.is_vendor,f.path,s.start", (symbol, symbol),
        ).fetchall()
        hits: list[dict[str, Any]] = []
        if op == "def":
            hits = [_row_hit(row, "def", "definition") for row in definitions]
            if len(definitions) > 1:
                coverage = "partial"
        elif op in {"refs", "callers"}:
            dst_ids = [row["id"] for row in definitions]
            if dst_ids:
                placeholders = ",".join("?" for _ in dst_ids)
                rows = conn.execute(
                    f"SELECT e.line,e.kind,e.source,e.confidence,f.path,f.is_test,COALESCE(s.name,e.dst_name) name,"
                    "COALESCE(s.start,e.line) start,COALESCE(s.end,e.line) end,COALESCE(s.sig,'') sig,COALESCE(s.snippet,'') snippet "
                    "FROM edges e JOIN files f ON f.id=e.file_id LEFT JOIN symbols s ON s.id=e.src_symbol_id "
                    f"WHERE e.dst_symbol_id IN ({placeholders}) OR e.dst_name=? ORDER BY (e.source='lsp') DESC,f.is_test,f.path,e.line",
                    (*dst_ids, symbol),
                ).fetchall()
            else:
                coverage = "partial"
                rows = conn.execute(
                    "SELECT e.line,e.kind,e.source,e.confidence,f.path,f.is_test,COALESCE(s.name,e.dst_name) name,"
                    "COALESCE(s.start,e.line) start,COALESCE(s.end,e.line) end,COALESCE(s.sig,'') sig,COALESCE(s.snippet,'') snippet "
                    "FROM edges e JOIN files f ON f.id=e.file_id LEFT JOIN symbols s ON s.id=e.src_symbol_id "
                    "WHERE e.dst_name=? ORDER BY (e.source='lsp') DESC,f.is_test,f.path,e.line", (symbol,),
                ).fetchall()
            hits = [
                _row_hit(
                    row, "call" if row["kind"] == "call" else "ref",
                    f"{row['source']} {row['kind']} of {symbol}",
                    float(row["confidence"]) if row["source"] == "lsp" else .9,
                )
                for row in rows
            ]
            if len(definitions) > 1:
                coverage = "partial"
        elif op == "callees":
            ids = [row["id"] for row in definitions]
            rows = [] if not ids else conn.execute(
                f"SELECT d.*,f.path,f.is_test FROM edges e JOIN symbols d ON d.id=e.dst_symbol_id JOIN files f ON f.id=d.file_id WHERE e.src_symbol_id IN ({','.join('?' for _ in ids)}) ORDER BY f.path,d.start",
                ids,
            ).fetchall()
            hits = [_row_hit(row, "call", f"called by {symbol}", .9) for row in rows]
            coverage = "partial"
        elif op in {"path", "impact"}:
            base = [_row_hit(row, "def", "definition") for row in definitions]
            callers = graph_query("callers", symbol, depth=max(1, depth - 1), budget_tokens=budget_tokens, start=start)
            hits = base + callers["hits"]
            coverage = "partial"
        else:
            raise ValueError(f"opération graph inconnue: {op}")
        unique: dict[tuple[str, int, str], dict[str, Any]] = {}
        for hit in hits:
            unique.setdefault((hit["path"], hit["start"], hit["kind"]), hit)
        return apply_budget(list(unique.values()), budget_tokens, started=started, coverage=coverage,
                            hint="résolution statique best-effort" if coverage == "partial" else None)
    finally:
        conn.close()

File: src/graph.rs
use std::collections::HashSet;
use std::path::Path;
use std::time::Instant;

use anyhow::{Result, bail};
use rusqlite::{Connection, params_from_iter};

use crate::config::find_ctx;
use crate::db::connect;
use crate::model::{Envelope, Hit};
use crate::search::apply_budget;

#[derive(Debug)]
struct Definition {
    id: i64,
    hit: Hit,
}

pub fn graph_query(
    operation: &str,
    symbol: &str,
    depth: usize,
    budget_tokens: usize,
    start: impl AsRef<Path>,
) -> Result<Envelope> {
    let started = Instant::now();
    let connection = connect(&find_ctx(start)?.join("index.sqlite"), false)?;
    let definitions = definitions(&connection, symbol)?;
    let mut coverage = "complete";
    let mut hint = None;
    let mut hits = match operation {
        "def" => definitions.iter().map(|item| item.hit.clone()).collect(),
        "refs" | "callers" => {
            if definitions.is_empty() || definitions.len() > 1 {
                coverage = "partial";
                hint = Some("résolution statique best-effort".to_owned());
            }
            references(&connection, symbol, &definitions)?
        }
        "callees" => {
            coverage = "partial";
            hint = Some("résolution statique best-effort".to_owned());
            callees(&connection, &definitions)?
        }
        "path" | "impact" => {
            coverage = "partial";
            hint = Some(format!(
                "résolution statique best-effort; profondeur demandée {depth}"
            ));
            let mut values = definitions
                .iter()
                .map(|item| item.hit.clone())
                .collect::<Vec<_>>();
            values.extend(references(&connection, symbol, &definitions)?);
            values
        }
        _ => bail!("opération graph inconnue: {operation}"),
    };
    if operation == "def" && definitions.len() > 1 {
        coverage = "partial";
        hint = Some("plusieurs définitions correspondent".to_owned());
    }
    let mut seen = HashSet::new();
    hits.retain(|hit| seen.insert((hit.path.clone(), hit.start, hit.kind.clone())));
    Ok(apply_budget(hits, budget_tokens, started, coverage, hint))
}

fn definitions(connection: &Connection, symbol: &str) -> Result<Vec<Definition>> {
    let mut statement = connection.prepare(
        "SELECT s.id,f.path,s.start,s.end,s.name,COALESCE(s.sig,''),
                COALESCE(s.snippet,''),f.is_test
         FROM symbols s JOIN files f ON f.id=s.file_id
         WHERE s.name=?1 OR s.qualname=?1
         ORDER BY f.is_test,f.is_vendor,f.path,s.start",
    )?;
    let rows = statement.query_map([symbol], |row| {
        Ok(Definition {
            id: row.get(0)?,
            hit: Hit {
                path: row.get(1)?,
                start: row.get::<_, i64>(2)?.max(1) as usize,
                end: row.get::<_, i64>(3)?.max(1) as usize,
                symbol: Some(row.get(4)?),
                kind: if row.get::<_, bool>(7)? {
                    "test"
                } else {
                    "def"
                }
                .to_owned(),
                sig: row.get(5)?,
                snippet: row.get(6)?,
                score: 1.0,
                why: "definition".to_owned(),
            },
        })
    })?;
    Ok(rows.collect::<rusqlite::Result<Vec<_>>>()?)
}

fn references(
    connection: &Connection,
    symbol: &str,
    definitions: &[Definition],
) -> Result<Vec<Hit>> {
    let ids = definitions.iter().map(|item| item.id).collect::<Vec<_>>();
    let (sql, values): (String, Vec<rusqlite::types::Value>) = if ids.is_empty() {
        (
            "SELECT e.line,e.kind,e.source,e.confidence,f.path,f.is_test,
                    COALESCE(s.name,e.dst_name),COALESCE(s.start,e.line),
                    COALESCE(s.end,e.line),COALESCE(s.sig,''),COALESCE(s.snippet,'')
             FROM edges e JOIN files f ON f.id=e.file_id
             LEFT JOIN symbols s ON s.id=e.src_symbol_id
             WHERE e.dst_name=?1
             ORDER BY (e.source='lsp') DESC,f.is_test,f.path,e.line"
                .to_owned(),
            vec![symbol.to_owned().into()],
        )
    } else {
        let placeholders = (1..=ids.len())
            .map(|index| format!("?{index}"))
            .collect::<Vec<_>>()
            .join(",");
        let symbol_parameter = ids.len() + 1;
        let sql = format!(
            "SELECT e.line,e.kind,e.source,e.confidence,f.path,f.is_test,
                    COALESCE(s.name,e.dst_name),COALESCE(s.start,e.line),
                    COALESCE(s.end,e.line),COALESCE(s.sig,''),COALESCE(s.snippet,'')
             FROM edges e JOIN files f ON f.id=e.file_id
             LEFT JOIN symbols s ON s.id=e.src_symbol_id
             WHERE e.dst_symbol_id IN ({placeholders}) OR e.dst_name=?{symbol_parameter}
             ORDER BY (e.source='lsp') DESC,f.is_test,f.path,e.line"
        );
        let mut values = ids
            .iter()
            .copied()
            .map(rusqlite::types::Value::Integer)
            .collect::<Vec<_>>();
        values.push(symbol.to_owned().into());
        (sql, values)
    };
    let mut statement = connection.prepare(&sql)?;
    let rows = statement.query_map(params_from_iter(values), |row| {
        let edge_kind = row.get::<_, String>(1)?;
        let source = row.get::<_, String>(2)?;
        Ok(Hit {
            path: row.get(4)?,
            start: row.get::<_, i64>(7)?.max(1) as usize,
            end: row.get::<_, i64>(8)?.max(1) as usize,
            symbol: Some(row.get(6)?),
            kind: if edge_kind == "call" { "call" } else { "ref" }.to_owned(),
            sig: row.get(9)?,
            snippet: row.get(10)?,
            score: if source == "lsp" { row.get(3)? } else { 0.9 },
            why: format!("{source} {edge_kind} of {symbol}"),
        })
    })?;
    Ok(rows.collect::<rusqlite::Result<Vec<_>>>()?)
}

fn callees(connection: &Connection, definitions: &[Definition]) -> Result<Vec<Hit>> {
    if definitions.is_empty() {
        return Ok(Vec::new());
    }
    let ids = definitions.iter().map(|item| item.id).collect::<Vec<_>>();
    let placeholders = (1..=ids.len())
        .map(|index| format!("?{index}"))
        .collect::<Vec<_>>()
        .join(",");
    let sql = format!(
        "SELECT d.name,d.start,d.end,COALESCE(d.sig,''),COALESCE(d.snippet,''),f.path
         FROM edges e JOIN symbols d ON d.id=e.dst_symbol_id
         JOIN files f ON f.id=d.file_id
         WHERE e.src_symbol_id IN ({placeholders})
         ORDER BY f.path,d.start"
    );
    let mut statement = connection.prepare(&sql)?;
    let rows = statement.query_map(params_from_iter(ids), |row| {
        Ok(Hit {
            path: row.get(5)?,
            start: row.get::<_, i64>(1)?.max(1) as usize,
            end: row.get::<_, i64>(2)?.max(1) as usize,
            symbol: Some(row.get(0)?),
            kind: "call".to_owned(),
            sig: row.get(3)?,
            snippet: row.get(4)?,
            score: 0.9,
            why: "callee".to_owned(),
        })
    })?;
    Ok(rows.collect::<rusqlite::Result<Vec<_>>>()?)
}
